### PR TITLE
Aarnq/long term audit logs

### LINKED ISSUE #847: Move fluentd s3 credentials to a helm chart.
**Improvement**

The fluentd s3 credentials is created by using kubectl. This should be moved to a helm chart as it is easy it when updating s3 credentials.

Currently created here:
https://github.com/elastisys/compliantkubernetes-apps/blob/v0.19.0/scripts/deploy-sc.sh#L22-L31

**Expected behavior**
Created and managed by helm.

### LINKED ISSUE #988: Fluentd: Update @id to match fluentd-wc.yaml
There is a `@id` which is still elasticsearch and should be changed to opensearch as it is in `fluentd-wc.yaml`

Correct:
https://github.com/elastisys/compliantkubernetes-apps/blob/main/helmfile/values/fluentd-wc.yaml.gotmpl#L238
Incorrect:
https://github.com/elastisys/compliantkubernetes-apps/blob/main/helmfile/values/fluentd-user.yaml.gotmpl#L220

### LINKED ISSUE #989: Drop support for GCS
We have not maintained or tested GCS for ck8s-apps for a long while and we should start considering dropping support for GCS.

### LINKED ISSUE #991: Fluentd: extract fluentd config files
At the moment we have multiple fluentd config files located in fluentd-user/wc. To make it more clear what config files we have and easier to see which one is changed during a PR. We should consider to extract these config files to separate files. 

Current config files [user/wc](https://github.com/elastisys/compliantkubernetes-apps/blob/main/helmfile/values/fluentd-user.yaml.gotmpl#L53)
- system.conf
- 10-monitoring.conf
- 10-kube-audit.conf
- containers.input.conf
- output.conf

This would also mean that we could share configs between fluentd deployments.

This should be done after we have restructured fluentd gotmpl files. 
https://github.com/elastisys/compliantkubernetes-apps/issues/990

### LINKED ISSUE #1379: Implement logging for HIPAA/PDL compliance
**Is your feature request related to a problem? Please describe.**

TODO

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Definition of done:**

### PR DESCRIPTION
**Pull request by aarnq**
**What this PR does / why we need it**:

We want long term audit logs to be able to comply with HIPAA and PDL so this changes quite a lot in how we deploy and run fluentd.

- Adds fluentd-aggregator in the workload cluster to aggregate audit logs and push them to object storage
- Switches the fluentd-forwarders in the service cluster to use fluentd-elasticsearch
- Configures fluentd with cascading templates to maximise reuse between helm releases

**Which issue this PR fixes**:
fixes #1379
kinda #992 they at least use the same for forwarding and aggregating
fixes #991
fixes #989
fixes #988 as far as I can tell
fixes #847

**Special notes for reviewers**:

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [ ] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [ ] is completely transparent, will not impact the workload in any way.
    - [x] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [x] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.
- Chart checklist (pick exactly one):
    - [x] I upgraded no Chart.
    - [ ] I upgraded a Chart and determined that no migration steps are needed.
    - [ ] I upgraded a Chart and added [migration steps](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration).

**Pipeline config** *(if applicable)*
If you change some config options (e.g. add/rename variable or change the default value) you may need to update the config used by the pipeline in `pipeline/config`.

### CODE REVIEW

File: helmfile/values/fluentd-configmap.yaml.gotmpl [hunk 1]
```diff
@@ -46,43 +46,126 @@ aggregator:
...
+          @type rewrite_tag_filter
+          <rule>
+            key message
+            pattern /AuditLog/
+            tag audit.${tag}
+          </rule>
+        </match>
+
+        # Set aside OpenSearch Master Audit events
+        <match kubernetes.var.log.containers.opensearch-master-**>
```
Comment (OlleLarsson) [inline helmfile/values/fluentd-configmap.yaml.gotmpl:86]: Is it only the opensearch master node that emits audit logs?

File: helmfile/values/fluentd/aggregator-workload-cluster.yaml.gotmpl [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        @label @AUDIT
+      </match>
+
+      # Relabel kubernetes
+      <match kubernetes.**>
+        @type relabel
+        @label @AUDIT
+      </match>
+
+    fluentd-filters.conf: |
```
Comment (OlleLarsson) [inline helmfile/values/fluentd/aggregator-workload-cluster.yaml.gotmpl:22]: Should this be empty?

File: helmfile/values/fluentd-configmap.yaml.gotmpl (hunk 1 above)
Comment (aarnq) [reply helmfile/values/fluentd-configmap.yaml.gotmpl:86]: I haven't yet checked.

File: helmfile/values/fluentd/aggregator-workload-cluster.yaml.gotmpl (hunk 2 above)
Comment (aarnq) [reply helmfile/values/fluentd/aggregator-workload-cluster.yaml.gotmpl:22]: No I will add so we can add filter through the config to capture AMS audit logs. I'm in the process of refactoring.

File: helmfile/values/fluentd/aggregator-workload-cluster.yaml.gotmpl [hunk 3]
```diff
@@ -0,0 +1,23 @@
+aggregator:
+  configMapFiles:
+    fluentd-labels.conf: |
+      # Relabel authlog
+      <match authlog.**>
+        @type relabel
+        @label @AUDIT
+      </match>
+
+      # Split kubeaudit
```
Comment (OlleLarsson) [inline helmfile/values/fluentd/aggregator-workload-cluster.yaml.gotmpl:10]: ?
```suggestion
      # Relable kubeaudit
```

**Comment by aarnq** [comment]
Should I go ahead with this https://github.com/elastisys/compliantkubernetes-apps/issues/989 while I'm at it?

**Comment by aarnq** [comment]
@OlleLarsson @robinAwallace it's sorta ready for review now. All features are there, I will clear out GCS support and do some clean up.

File: migration/v0.28.x-v0.29.x/upgrade-apps.md [hunk 4]
```diff
@@ -10,6 +10,23 @@
     bin/ck8s init
     ```
 
+1. Run migration script `reconfigure-fluentd.sh`
+
+    *Optional:*
+    Fluentd can now collect audit logs, enable it by setting `fluentd.audit.enable: true`.
+    Make sure to create the bucket as set by `objectStore.buckets.audit` to store those logs.
```
Comment (robinAwallace) [inline migration/v0.28.x-v0.29.x/upgrade-apps.md:17]:
```suggestion
    Make sure to create the bucket as set by `objectStore.buckets.audit` to store those logs.
    
    ```
    scripts/S3/entry.sh --s3cfg $CK8S_CONFIG_PATH/.state/s3cfg.ini create
    ```
```

File: helmfile/values/networkpolicy/service-cluster.yaml.gotmpl [hunk 5]
```diff
@@ -22,6 +22,19 @@ global:
   externalLoadBalancer: {{ .Values.networkPolicies.global.externalLoadBalancer }}
   ingressUsingHostNetwork: {{ .Values.ingressNginx.controller.useHostPort }}
 
+  # TODO: Add Swift
```
Comment (robinAwallace) [inline helmfile/values/networkpolicy/service-cluster.yaml.gotmpl:25]: TODO part of another PR?

File: config/config/sc-config.yaml [hunk 6]
```diff
@@ -976,52 +976,53 @@ opensearch:
 
 fluentd:
   enabled: true
```
Comment (robinAwallace) [inline config/config/sc-config.yaml:978]: Move to common

File: helmfile/charts/log-manager/files/compaction.sh [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+s3_put_chunk() {
+  s3cmd --config "$S3_CONFIG" put --no-preserve "$2" "s3://$S3_BUCKET/$S3_PREFIX/$1/" > /dev/null
+}
+
+s3_rm_chunks() {
+  xargs -n1000 s3cmd --config "$S3_CONFIG" rm < "$1" > /dev/null
+}
+
+merge_chunks() {
+  INDICES="$(s3_list_indices "$1")"
```
Comment (robinAwallace) [inline helmfile/charts/log-manager/files/compaction.sh:46]:
```suggestion
merge_chunks() {
  DAY=$1
  INDICES="$(s3_list_indices "${DAY}")"
```

File: helmfile/charts/log-manager/files/retention.sh [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+LIMIT="$(date --utc --date="$RETAIN_DAYS days ago" '+%Y%m%d')"
+
+TMPFILE="/tmp/lm.idx"
+
+s3_list_days() {
+  s3cmd --config "$S3_CONFIG" ls "s3://$S3_BUCKET/$S3_PREFIX/" | grep 'DIR' | awk '{print $2}' | sed "s#s3://$S3_BUCKET/$S3_PREFIX/##" | sed 's#/$##'
+}
+
+s3_list_chunks() {
+  s3cmd --config "$S3_CONFIG" ls -r "s3://$S3_BUCKET/$S3_PREFIX/$1/" | grep '\.gz\|\.zst' | awk '{print $4}'
```
Comment (robinAwallace) [inline helmfile/charts/log-manager/files/retention.sh:21]:
```suggestion
s3_list_chunks() {
  DAY=$1
  s3cmd --config "$S3_CONFIG" ls -r "s3://$S3_BUCKET/$S3_PREFIX/$DAY/" | grep '\.gz\|\.zst' | awk '{print $4}'
```

File: helmfile/charts/log-manager/values.yaml [hunk 9]
```diff
@@ -0,0 +1,57 @@
+fullnameOverride: ""
+nameOverride: ""
+
+imagePullSecrets: []
+image:
+  pullPolicy: IfNotPresent
+  repository: harbor.aarnq.dev-ck8s.com/ck8s/log-manager
```
Comment (robinAwallace) [inline helmfile/charts/log-manager/values.yaml:7]: Point to a public repo with the image?

File: helmfile/02-networkpolicies.yaml [hunk 10]
```diff
@@ -34,7 +34,7 @@ releases:
   chart: ./charts/networkpolicy-common
```
Comment (robinAwallace) [inline helmfile/02-networkpolicies.yaml:34]: Do the same thing for the chart?
```suggestion
  chart: ./charts/networkpolicy/common
```

**Review by robinAwallace — COMMENTED** [review]
Just minor things. Could be good with a meeting so you can give a short overview of the changes?

File: config/config/sc-config.yaml (hunk 6 above)
Comment (aarnq) [reply config/config/sc-config.yaml:978]: It'll break the dev flavour :disappointed:

File: helmfile/values/networkpolicy/service-cluster.yaml.gotmpl (hunk 5 above)
Comment (aarnq) [reply helmfile/values/networkpolicy/service-cluster.yaml.gotmpl:25]: Yeah, it's for the great refactoring of netpols.

File: helmfile/charts/log-manager/values.yaml (hunk 9 above)
Comment (aarnq) [reply helmfile/charts/log-manager/values.yaml:7]: Forgot about pushing that one :sweat_smile:

File: WIP-CHANGELOG.md [hunk 11]
```diff
@@ -23,3 +26,10 @@
 - Allow the creation of arbitrary network-policy in wc
 - Network polices
   - Added missing network policy for rook-ceph-csi-detect-version for rook-ceph v1.10
+- Fluentd in both SC and WC now use the fluentd-elasticsearch chart for forwarding, and the fluentd chart for aggregating
```
Comment (OlleLarsson) [inline WIP-CHANGELOG.md:29]:
```suggestion
- Fluentd in both SC and WC now use the `fluentd-elasticsearch` chart for forwarding, and the `fluentd` chart for aggregating
```

File: config/config/common-config.yaml [hunk 12]
```diff
@@ -156,6 +157,65 @@ falco:
     affinity: {}
     nodeSelector: {}
 
+fluentd:
+  audit:
+    enabled: false
+    # Add filter stages to capture audit logs, add the audit tag prefix
+    filters: ""
```
Comment (OlleLarsson) [inline config/config/common-config.yaml:164]: ?
```suggestion
    filters: |-
```

Comment (aarnq) [reply config/config/common-config.yaml:164]: It's a string nonetheless.

File: helmfile/charts/log-manager/templates/secret.yaml [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+  name: {{ include "log-manager.fullname" . }}-s3-credentials
+  labels:
+    {{- include "log-manager.labels" . | nindent 4 }}
+stringData:
+  s3cfg.ini: |
+    host_base         = {{ .Values.s3.regionEndpoint }}
+    {{- if .Values.s3.forcePathStyle }}
+    host_bucket       = {{ .Values.s3.regionEndpoint }}
+    {{- else }}
+    host_bucket       = %(bucket)s.{{ .Values.s3.regionEndpoint }}
```
Comment (OlleLarsson) [inline helmfile/charts/log-manager/templates/secret.yaml:14]: Does this work if region endpoint contains https/http? 
```suggestion
    host_bucket       = %(bucket)s.{{ .Values.s3.regionEndpoint }}
```

Comment (aarnq) [reply helmfile/charts/log-manager/templates/secret.yaml:14]: Good catch, it won't.

**Comment by OlleLarsson** [comment]
Remove https://github.com/elastisys/compliantkubernetes-apps/blob/d0de9553942183290a900eca4f903cd4414784bb/config/config/sc-config.yaml#L909-L911 as the audit logs are no longer sent to an index.

**Comment by OlleLarsson** [comment]
Remove https://github.com/elastisys/compliantkubernetes-apps/blob/d0de9553942183290a900eca4f903cd4414784bb/config/config/sc-config.yaml#L909-L911 as the audit longs are no longer stored in indices.
And same for https://github.com/elastisys/compliantkubernetes-apps/blob/d0de9553942183290a900eca4f903cd4414784bb/config/config/sc-config.yaml#L933-L934

**Review by robinAwallace — APPROVED** [review]
To me this looks good :+1: :partying_face:

File: WIP-CHANGELOG.md [hunk 14]
```diff
@@ -27,3 +29,10 @@
 - Allow the creation of arbitrary network-policy in wc
 - Network polices
   - Added missing network policy for rook-ceph-csi-detect-version for rook-ceph v1.10
+- Fluentd in both SC and WC now use the `fluentd-elasticsearch` chart for forwarding, and the `fluentd` chart for aggregating
+  - This have changed the deployment for Fluentd considerably and users must ensure that their custom filter continues to work as expected
```
Comment (crssnd) [inline WIP-CHANGELOG.md:33]: maybe mention this in the release notes also, to be more visible

**Review by crssnd — APPROVED** [review]

**Pull request merged**